geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/SlotAdvisor.java

```diff
@@ -0,0 +1,230 @@
+
+public class SlotAdvisor {
+
+  private static final Logger logger = LogService.getLogger();
+
+  /**
+   * Mapping buckets to slot information
+   */
+  private final List<MemberBucketSlot> memberBucketSlots;
+
+  /**
+   * Cache of member ids to member IP address and redis listening port
+   */
+  private final Map<String, Pair<String, Integer>> hostPorts = new HashMap<>();
+  private final PartitionedRegion dataRegion;
+  private final Region<String, Object> configRegion;
+
+  SlotAdvisor(Region<RedisKey, RedisData> dataRegion, Region<String, Object> configRegion) {
+    this.dataRegion = (PartitionedRegion) dataRegion;
+    this.configRegion = configRegion;
+    memberBucketSlots = new ArrayList<>(RegionProvider.REDIS_REGION_BUCKETS);
+    for (int i = 0; i < REDIS_REGION_BUCKETS; i++) {
+      memberBucketSlots.add(null);
+    }
+  }
+
+  public boolean isLocal(RedisKey key) {
+    return dataRegion.getRegionAdvisor().getBucket(key.getBucketId()).getBucketAdvisor()
+        .isPrimary();
+  }
+
+  public Pair<String, Integer> getHostAndPortForKey(RedisKey key) {
+    int bucketId = key.getBucketId();
+    MemberBucketSlot mbs = updateBucketDetails(bucketId);
+
+    return Pair.of(mbs.getPrimaryIpAddress(), mbs.getPrimaryPort());
+  }
+
+  public Map<String, List<Integer>> getMemberBuckets() {
+    initializeBucketsIfNecessary();
+
+    Map<String, List<Integer>> memberBuckets = new HashMap<>();
+    for (int bucketId = 0; bucketId < REDIS_REGION_BUCKETS; bucketId++) {
+      String memberId =
+          dataRegion.getRegionAdvisor().getBucketAdvisor(bucketId).getPrimary().getUniqueId();
+      memberBuckets.computeIfAbsent(memberId, k -> new ArrayList<>()).add(bucketId);
+    }
+
+    return memberBuckets;
+  }
+
+  public synchronized List<MemberBucketSlot> getBucketSlots() {
+    initializeBucketsIfNecessary();
+
+    for (int bucketId = 0; bucketId < REDIS_REGION_BUCKETS; bucketId++) {
+      updateBucketDetails(bucketId);
+    }
+
+    return Collections.unmodifiableList(memberBucketSlots);
+  }
+
+  private synchronized MemberBucketSlot updateBucketDetails(int bucketId) {
+    MemberBucketSlot mbs = null;
+    try {
+      Pair<String, Integer> hostPort = getHostPort(bucketId);
+
+      mbs = memberBucketSlots.get(bucketId);
+      if (mbs == null) {
+        mbs = new MemberBucketSlot(bucketId, hostPort.getLeft(), hostPort.getRight());
+        memberBucketSlots.set(bucketId, mbs);
+      } else {
+        mbs.setPrimaryIpAddress(hostPort.getLeft());
+        mbs.setPrimaryPort(hostPort.getRight());
+      }
+    } catch (Exception ex) {
+      logger.error("Unable to update bucket detail for bucketId: {}", bucketId, ex);
+    }
+
+    return mbs;
+  }
+
+  private void initializeBucketsIfNecessary() {
+    if (dataRegion.getDataStore() != null &&
+        dataRegion.getDataStore().getAllLocalBucketIds().isEmpty()) {
+      PartitionRegionHelper.assignBucketsToPartitions(dataRegion);
+    }
+  }
+
+  @SuppressWarnings("unchecked")
```

Comment:
Why do we need a suppress warnings?

Reply:
I've pushed this down to the exact line it suppresses.

Reply:
But now it's back again because ... 🤷